src/silo_api/database_directory_watcher.cpp

```diff
@@ -0,0 +1,126 @@
+#include "silo_api/database_directory_watcher.h"
+
+#include <ranges>
+
+#include <boost/algorithm/string/join.hpp>
+
+silo_api::DatabaseDirectoryWatcher::DatabaseDirectoryWatcher(
+   const std::filesystem::path& path,
+   DatabaseMutex& database_mutex
+)
+    : path(path),
+      database_mutex(database_mutex),
+      timer(0, 2000) {
+   timer.start(Poco::TimerCallback<DatabaseDirectoryWatcher>(
+      *this, &DatabaseDirectoryWatcher::checkDirectoryForData
+   ));
+}
+
+namespace {
+
+std::optional<silo::DataVersion> checkValidDataSource(const std::filesystem::path& path) {
+   if (!std::filesystem::is_directory(path)) {
+      SPDLOG_INFO("Skipping {} because it is not a directory", path.string());
+      return std::nullopt;
+   }
+   auto data_version = silo::DataVersion::fromString(path.filename().string());
+   if (data_version == std::nullopt) {
+      SPDLOG_INFO("Skipping {}. Its name is not a valid data version.", path.string());
+      return std::nullopt;
+   }
+   auto data_version_filename = path / "data_version.silo";
+   if (!std::filesystem::is_regular_file(data_version_filename)) {
+      SPDLOG_INFO(
+         "Skipping {}. it does not contain the data version file {}, which "
+         "confirms a finished and valid data source",
+         path.string(),
+         data_version_filename.string()
+      );
+      return std::nullopt;
+   }
+   std::ifstream data_version_file(data_version_filename);
+   if (!data_version_file) {
+      SPDLOG_INFO("Skipping {}. The data version file could not be opened", path.string());
+      return std::nullopt;
+   }
+   std::string data_version_in_file_string;
+   data_version_file >> data_version_in_file_string;
+   const auto data_version_in_file = silo::DataVersion::fromString(data_version_in_file_string);
+   if (data_version_in_file == std::nullopt) {
+      SPDLOG_INFO(
+         "Skipping {}. The data version in data_version.silo could not be parsed", path.string()
+      );
+      return std::nullopt;
+   }
+   if (data_version_in_file != data_version) {
+      SPDLOG_INFO(
+         "Skipping {}. The data version in data_version.silo is not equal to the directory name",
+         path.string()
+      );
+      return std::nullopt;
+   }
+   return data_version.value();
```

Comment:
I think this yields the same but saves unnecessary conversions:
```suggestion
   return data_version;
```